src/iceberg/literal.h

```diff
@@ -0,0 +1,138 @@
+                   float,                    // for float
+                   double,                   // for double
+                   std::string,              // for string
+                   std::vector<uint8_t>,     // for binary, fixed
+                   std::array<uint8_t, 16>,  // for uuid and decimal
+                   BelowMin, AboveMax>;
+
+ public:
+  /// Factory methods for primitive types
+  static PrimitiveLiteral Boolean(bool value);
+  static PrimitiveLiteral Int(int32_t value);
+  static PrimitiveLiteral Long(int64_t value);
+  static PrimitiveLiteral Float(float value);
+  static PrimitiveLiteral Double(double value);
+  static PrimitiveLiteral String(std::string value);
+  static PrimitiveLiteral Binary(std::vector<uint8_t> value);
+
+  /// Create iceberg literal from bytes.
+  ///
+  /// See [this spec](https://iceberg.apache.org/spec/#binary-single-value-serialization)
+  /// for reference.
+  static Result<PrimitiveLiteral> Deserialize(std::span<const uint8_t> data);
+
+  /// Serialize iceberg literal to bytes.
+  ///
+  /// See [this spec](https://iceberg.apache.org/spec/#binary-single-value-serialization)
+  /// for reference.
+  Result<std::vector<uint8_t>> Serialize() const;
+
+  /// Get the Iceberg Type of the literal.
+  const std::shared_ptr<PrimitiveType>& type() const;
+
+  /// Converts this literal to a literal of the given type.
+  ///
+  /// When a predicate is bound to a concrete data column, literals are converted to match
+  /// the bound column's type. This conversion process is more narrow than a cast and is
+  /// only intended for cases where substituting one type is a common mistake (e.g. 34
+  /// instead of 34L) or where this API avoids requiring a concrete class (e.g., dates).
+  ///
+  /// If conversion to a target type is not supported, this method returns an error.
+  ///
+  /// This method may return BelowMin or AboveMax when the target type is not as wide as
+  /// the original type. These values indicate that the containing predicate can be
+  /// simplified. For example, std::numeric_limits<int>::max()+1 converted to an int will
+  /// result in AboveMax and can simplify a < std::numeric_limits<int>::max()+1 to always
+  /// true.
+  ///
+  /// \param target_type A primitive PrimitiveType
+  /// \return A Result containing a literal of the given type or an error if conversion
+  /// was not valid
+  Result<PrimitiveLiteral> CastTo(
+      const std::shared_ptr<PrimitiveType>& target_type) const;
+
+  /// Compare two PrimitiveLiterals. Both literals must have the same type
+  /// and should not be AboveMax or BelowMin.
+  std::partial_ordering operator<=>(const PrimitiveLiteral& other) const;
+
+  bool isAboveMax() const;
+  bool isBelowMin() const;
+
+  std::string ToString() const;
+
+ private:
+  PrimitiveLiteral(PrimitiveLiteralValue value, std::shared_ptr<PrimitiveType> type);
+
+  static PrimitiveLiteral BelowMinLiteral(std::shared_ptr<PrimitiveType> type);
```

Comment:
I don't think we need a type here, right? If your intention is to make them comparable, I think we can add `ToAboveMax()` and `ToBelowMin()` to replace these two static functions to keep the original type.

Reply:
> I don't think we need a type here, right?

`BelowMin` and `AboveMax` should be typed, a BelowMin for int32 and other types have different meaning

Reply:
sample: `Literal(type:int64, value: int32_t::max() + 1)` cast to int32_t becomes `Literal(type:int32, value: AboveMax)`